src/goto-programs/remove_returns.cpp

```diff
@@ -46,74 +51,91 @@ class remove_returnst
   void undo_function_calls(
     goto_functionst &goto_functions,
     goto_programt &goto_program);
+
+  symbol_exprt get_or_create_return_value_symbol(const irep_idt &function_id);
 };
 
+symbol_exprt remove_returnst::get_or_create_return_value_symbol(
+  const irep_idt &function_id)
+{
+  const irep_idt symbol_name = id2string(function_id) + RETURN_VALUE_SUFFIX;
+  const symbolt *existing_symbol = symbol_table.lookup(symbol_name);
+  if(existing_symbol != nullptr)
+    return existing_symbol->symbol_expr();
+
+  const symbolt &function_symbol = symbol_table.lookup_ref(function_id);
+  const typet &return_type = to_code_type(function_symbol.type).return_type();
+
+  if(return_type == empty_typet())
+    return symbol_exprt();
+
+  auxiliary_symbolt new_symbol;
+  new_symbol.is_static_lifetime = true;
+  new_symbol.module = function_symbol.module;
+  new_symbol.base_name =
+    id2string(function_symbol.base_name) + RETURN_VALUE_SUFFIX;
+  new_symbol.name = symbol_name;
+  new_symbol.mode = function_symbol.mode;
+  // If we're creating this for the first time, the target function cannot have
+  // been remove_return'd yet, so this will still be the "true" return type:
+  new_symbol.type = return_type;
+  // Return-value symbols will always be written before they are read, so there
+  // is no need for __CPROVER_initialize to do anything:
+  new_symbol.type.set(ID_C_no_initialization_required, true);
```

Comment:
I know I should have commented on this days ago, but anyway: I like the idea, but would feel the `.value` field is a better place for that than the `.type`? It's not a property that is tied to the type, is it? I'm also saying this as the type continues to exist (that very same `irept` can be shared across all the instantiations), while the value will be overwritten anyway.

Reply:
Neither is ideal -- really I want a `map<irep_idt, irept> attributes` defined for each symbol and GOTO function, but that's a breaking change. I chose the type because I thought it less likely to get copied into a place where the annotation no longer makes sense. I also wanted this to work with symbols with value `nil`, and nil-with-annotation-comment seemed dubious if anyone uses `full_eq` to compare them. However I can switch it if you prefer.

Reply:
Yes, one of the variants that are a breaking change is the right way to go. With that agreed, I'd think that `.value` is a better place for it isn't getting duplicated during symbolic execution, but for sure it's not the perfect place either. I'll leave it to you to decide, certainly not a blocker in any way.